File: OJ/LeetCode/120. Triangle.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        for(int x=triangle.size()-2;x>=0;x--){
            List<Integer> level = triangle.get(x);
            for(int y=0;y<=x;y++){
                level.set(y, triangle.get(x).get(y) + (triangle.get(x+1).get(y) < triangle.get(x+1).get(y+1) ? triangle.get(x+1).get(y) : triangle.get(x+1).get(y+1)));
            }
            triangle.set(x, level);
        }
        return triangle.get(0).get(0);
    }
}

File: OJ/LeetCode/120. Triangle.py
# 观察得，每一行的第i个点只能转移到下一行的第i个或者第i+1个点
# 然后运用dp即可
# 反过来dp会更快
class Solution:
    def minimumTotal(self, triangle):
        if len(triangle) == 1:
            return triangle[0][0]

        for level_n in range(len(triangle)-2, -1,-1):
            for n in range(len(triangle[level_n])):
                triangle[level_n][n] += min(triangle[level_n+1][n], triangle[level_n+1][n+1])

        return triangle[0][0]        
        